### ISSUE TITLE
Composer report the real php ext version instead of the one in composer.json.config.platform

### ISSUE THREAD

**Issue by lyrixx** [composer/composer#9876]
My `composer.json`:

```json
{
    "require": {
        "php": ">=7.4",
        "ext-mongodb": "*",
        "doctrine/mongodb-odm-bundle": "^4.3"
    },
    "config": {
        "sort-packages": true,
        "platform": {
            "ext-mongo": "1.6.14",
            "ext-mongodb": "1.3.4"
        }
    }
}

```

Output of `composer diagnose`:

```
/app $ composer diagnose
Checking composer.json: WARNING
doctrine/cache is required both in require and require-dev, this can lead to unexpected behavior
require.doctrine/common : exact version constraints (2.12.0) should be avoided if the package follows semantic versioning
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: FAIL
Missing pubkey for tags verification
Missing pubkey for dev verification
Run composer self-update --update-keys to set them up
Checking composer version: OK
Composer version: 2.0.13
PHP version: 7.4.7
PHP binary path: /usr/local/bin/php
OpenSSL version: OpenSSL 1.1.1g  21 Apr 2020
cURL version: 7.76.1 libz 1.2.11 ssl OpenSSL/1.1.1k
zip: extension present, unzip present

```

When I run this command: 

```
composer install -n 
```

```
Step 15/15 : RUN composer install -n
 ---> Running in 20239068e53e
No lock file found. Updating dependencies instead of installing from lock file. Use composer update over composer install if you do not have a lock file.
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - doctrine/mongodb-odm[2.0.0, ..., 2.2.0] require ext-mongodb ^1.5 -> it has the wrong version (1.7.4) installed. Install or enable PHP's mongodb extension.
    - doctrine/mongodb-odm-bundle 4.3.0 requires doctrine/mongodb-odm ^2.0.0 -> satisfiable by doctrine/mongodb-odm[2.0.0, ..., 2.2.0].
    - Root composer.json requires doctrine/mongodb-odm-bundle ^4.3 -> satisfiable by doctrine/mongodb-odm-bundle[4.3.0].

To enable extensions, verify that they are enabled in your .ini files:
    - 
    - /usr/local/etc/php/conf.d/0-apc.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-amqp.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-bcmath.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-intl.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-mongodb.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-mysqli.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-opcache.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-pcntl.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-sockets.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-sodium.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-xsl.ini
    - /usr/local/etc/php/conf.d/docker-php-ext-zip.ini
You can also run `php --ini` inside terminal to see which files are used by PHP in CLI mode.
```

And I expected this to happen:

It should says:
```diff
- it has the wrong version (1.7.4)
+ it has the wrong version (1.3.4)
```

---

**TL;DR** composer use the config.platform to solve the dependencies, but when reporting error it use the real extension installed. I had hard time to understand what was going on. I think composer could use a better error message to highlight the issue

I created a [reproducer with docker](https://gist.github.com/lyrixx/0076cbf8e11123788c59e9b131834518), I hope this will help

**Comment by stof** [comment]
The problem renderer already has the necessary logic for the `php` package to take the platform overrides into account: https://github.com/composer/composer/blob/0021758e0bcf964237a56cb22e2f0309118b6a3f/src/Composer/DependencyResolver/Problem.php#L183-L194
but for extensions, it directly uses the PHP runtime instead of using the PlatformRepository: https://github.com/composer/composer/blob/0021758e0bcf964237a56cb22e2f0309118b6a3f/src/Composer/DependencyResolver/Problem.php#L215-L216

So the fix would be to make it use the PlatformRepository logic for extensions too.

**Comment by stof** [comment]
@lyrixx that's what the logic based on packages already does (try configuring `platform.php` and triggering a conflict on the php package)
